### PR TITLE
feat: wysiwyg table move cursor

### PR DESCRIPTION
**Pull request by seonim-ryu**
### Please check if the PR fulfills these requirements
- [x] It's the right issue type on the title
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [ ] It does not introduce a breaking change or has a description of the breaking change

## Feature

In version 2, if there is no content before or after the table, a blank line is added or the table is deleted through the **cursor**. In version 3, the previous method cannot be used because of the **default behavior of ProseMirror that the cursor (`TextSelection`) cannot be added behind the table**, so the cursor operation method is improved. Depending on the key type and direction, it operates as follows.

### Up and down arrow key operation

* When there is no content before/after the table: Add an empty line (`paragraph`)

up-down-empty

* When there is content before/after the table: Move the cursor

up-down

### Tab key operation

* When there is no content before/after the table: Add an empty line (`paragraph`)

tab-empty

* When there is content before/after the table: Move the cursor

tab

### Left and right arrow key operation

* When pressing the `left` arrow key on the first cell of the table and the first cursor: Select a table (`NodeSelection`)
* When pressing the `right` arrow key on the last cell of the table or the last cursor: Select a table (`NodeSelection`)

left-right

In the case of the table removal operation, the table is removed through the context menu by default. When the table is removed with the `delete` key, it is removed only when the table is selected through the arrow keys.

### Table remove

* Remove with context menu

remove-table

* Select a table with the arrow keys and remove it by pressing the `delete` key.

left-right-delete

## TODO

* [ ] Change border color when selecting table

---
Thank you for your contribution to TOAST UI product. 🎉 😘 ✨

### CODE REVIEW

File: apps/editor/src/wysiwyg/nodes/table.ts [hunk 1]
```diff
@@ -285,20 +293,68 @@ export class Table extends NodeSchema {
...
+          newTr = addParagraphBeforeTable(tr, cellsInfo, schema);
+        } else if (cursorInTableEnd) {
+          newTr = addParagraphAfterTable(tr, cellsInfo, schema);
+        }
+
+        if (newTr) {
+          dispatch!(newTr);
+
+          return true;
+        }
```
Comment (js87zz) [inline apps/editor/src/wysiwyg/nodes/table.ts:320]: 별도 함수로 분리하는건 가독성이 더 떨어질 것 같고 주석으로 동작 설명을 구분하면 좋을 것 같습니다!

File: apps/editor/src/wysiwyg/nodes/table.ts [hunk 2]
```diff
@@ -285,20 +293,68 @@ export class Table extends NodeSchema {
...
+
+      return false;
+    };
+  }
+
+  private moveInCell(direction: CursorDirection): EditorCommand {
+    return () => (state, dispatch, view) => {
+      const { selection, tr } = state;
+
+      if (view?.endOfTextblock(direction)) {
```
Comment (js87zz) [inline apps/editor/src/wysiwyg/nodes/table.ts:340]: view 가 없는 경우가 없을 것 같은데 non-nullable로 처리해도 되지 않을까여?

File: apps/editor/src/wysiwyg/nodes/table.ts [hunk 3]
```diff
@@ -358,13 +414,20 @@ export class Table extends NodeSchema {
...
     return {
-      Tab: this.moveToCell('left')(),
-      'Shift-Tab': this.moveToCell('right')(),
+      Tab: this.moveToCell('right')(),
+      'Shift-Tab': this.moveToCell('left')(),
+
+      ArrowLeft: moveLeftInCellCommand,
+      ArrowRight: moveRightInCellCommand,
+      'Shift-ArrowLeft': moveLeftInCellCommand,
+      'Shift-ArrowRight': moveRightInCellCommand,
```
Comment (js87zz) [inline apps/editor/src/wysiwyg/nodes/table.ts:430]: 결국 일반 방향키랑 같은 역할 같은데 따로 제공할 필요가 있을까요?

File: apps/editor/src/wysiwyg/command/table.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import { Selection, Transaction } from 'prosemirror-state';
+
+import { createParagraph, createTextSelection } from '@/helper/manipulation';
+import { CellInfo } from '@/wysiwyg/helper/table';
+
+export type CursorDirection = 'left' | 'right';
+
+export type CellDirection = CursorDirection | 'up' | 'down';
+
+export function isCursorInTableStart(direction: CellDirection, [rowIndex, columnIndex]: number[]) {
```
Comment (js87zz) [inline apps/editor/src/wysiwyg/command/table.ts:11]: 다음 진행 방향에 따라 커서 위치가 테이블을 벗어난다는 의미이니, 
`canBeOutOfTableStart`, `canBeOutOfTableEnd`이런건 어떨까요?

File: apps/editor/src/wysiwyg/command/table.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+export function addParagraphBeforeTable(tr: Transaction, cellsInfo: CellInfo[][], schema: Schema) {
+  // 3 is position value of <table><thead><tr>
+  const tableStartPos = tr.doc.resolve(cellsInfo[0][0].offset - 3);
+
+  if (!tableStartPos.nodeBefore) {
+    const tableStartOffset = tableStartPos.pos;
+
+    tr.replaceWith(tableStartOffset, tableStartOffset, createParagraph(schema));
+
+    return tr.setSelection(createTextSelection(tr, tableStartOffset + 1));
```
Comment (js87zz) [inline apps/editor/src/wysiwyg/command/table.ts:42]: 이 코드는 공통으로 같이 써도 될것 같습니다~!

File: apps/editor/src/wysiwyg/nodes/table.ts [hunk 6]
```diff
@@ -285,20 +293,68 @@ export class Table extends NodeSchema {
...
+
+          return true;
+        }
+      }
+
+      return false;
+    };
+  }
+
+  private moveInCell(direction: CursorDirection): EditorCommand {
```
Comment (js87zz) [inline apps/editor/src/wysiwyg/nodes/table.ts:336]: 키맵들 관련해서 테스트도 추가되어야겠네요~!

**Review by js87zz — APPROVED** [review]
@seonim-ryu 
깔끔하게 처리되었네요ㅎㅎ 저는 위 아래 커서 이동 시에도 노드 셀렉션이 되는 걸로 생각했는데 동작을 보니까 좌, 우 이동시 노드 셀렉션 잡히는게 더 명확해서 좋네요. 고생많으셨습니다!!

**Review by lja1018 — APPROVED** [review]
리뷰완료
고생하셨습니다!

File: apps/editor/src/wysiwyg/nodes/table.ts (hunk 3 above)
Comment (seonim-ryu) [reply apps/editor/src/wysiwyg/nodes/table.ts:430]: 이건 셀렉션 동작이랑 관련된 키맵인데 제가 착각했네요 제거할게요~

File: apps/editor/src/wysiwyg/nodes/table.ts (hunk 6 above)
Comment (seonim-ryu) [reply apps/editor/src/wysiwyg/nodes/table.ts]: > https://github.com/nhn/tui.editor/pull/1374/commits/81d45f6993e4254ecfcc4f57a11863230ab6f553

**Pull request merged**